File: backend/src/models/preferences.rs
//! User preferences and status models

use std::time::SystemTime;

use chrono::{DateTime, NaiveTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

/// User status (displayed to other users)
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UserStatus {
    pub presence: Option<String>, // 'online', 'away', 'dnd', 'offline'
    #[serde(default)]
    pub manual: bool,
    #[serde(default = "SystemTime::now")]
    pub last_activity: SystemTime,
    pub text: Option<String>,
    pub emoji: Option<String>,
    pub expires_at: Option<DateTime<Utc>>,
}

/// User preferences from database
#[derive(Debug, Clone, Serialize, FromRow)]
pub struct UserPreferences {
    pub user_id: Uuid,

    // Notification preferences
    pub notify_desktop: String,
    pub notify_push: String,
    pub notify_email: String,
    pub notify_sounds: bool,

    // DND
    pub dnd_enabled: bool,
    pub dnd_start_time: Option<NaiveTime>,
    pub dnd_end_time: Option<NaiveTime>,
    pub dnd_days: String,

    // Display
    pub message_display: String,
    pub sidebar_behavior: String,
    pub time_format: String,

    // Keywords
    pub mention_keywords: Option<Vec<String>>,

    // Display settings (S7)
    pub collapsed_reply_threads: Option<bool>,
    pub use_military_time: Option<bool>,
    pub teammate_name_display: Option<String>,
    pub availability_status_visible: Option<bool>,
    pub show_last_active_time: Option<bool>,
    pub timezone: Option<String>,
    pub link_previews_enabled: Option<bool>,
    pub image_previews_enabled: Option<bool>,
    pub click_to_reply: Option<bool>,
    pub channel_display_mode: Option<String>,
    pub quick_reactions_enabled: Option<bool>,
    pub emoji_picker_enabled: Option<bool>,
    pub language: Option<String>,

    // Sidebar settings (S6)
    pub group_unread_channels: Option<String>,
    pub limit_visible_dms_gms: Option<String>,

    // Advanced settings (S5)
    pub send_on_ctrl_enter: Option<bool>,
    pub enable_post_formatting: Option<bool>,
    pub enable_join_leave_messages: Option<bool>,
    pub enable_performance_debugging: Option<bool>,
    pub unread_scroll_position: Option<String>,
    pub sync_drafts: Option<bool>,

    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// DTO for updating user status
#[derive(Debug, Clone, Deserialize)]
pub struct UpdateStatus {
    pub presence: Option<String>,
    pub text: Option<String>,
    pub emoji: Option<String>,
    #[serde(default)]
    pub duration_minutes: Option<i32>,
}

/// DTO for updating preferences
#[derive(Debug, Clone, Deserialize)]
pub struct UpdatePreferences {
    // Notifications
    pub notify_desktop: Option<String>,
    pub notify_push: Option<String>,
    pub notify_email: Option<String>,
    pub notify_sounds: Option<bool>,

    // DND
    pub dnd_enabled: Option<bool>,
    pub dnd_start_time: Option<String>,
    pub dnd_end_time: Option<String>,
    pub dnd_days: Option<String>,

    // Display
    pub message_display: Option<String>,
    pub sidebar_behavior: Option<String>,
    pub time_format: Option<String>,

    // Keywords
    pub mention_keywords: Option<Vec<String>>,

    // Display settings (S7)
    pub collapsed_reply_threads: Option<bool>,
    pub use_military_time: Option<bool>,
    pub teammate_name_display: Option<String>,
    pub availability_status_visible: Option<bool>,
    pub show_last_active_time: Option<bool>,
    pub timezone: Option<String>,
    pub link_previews_enabled: Option<bool>,
    pub image_previews_enabled: Option<bool>,
    pub click_to_reply: Option<bool>,
    pub channel_display_mode: Option<String>,
    pub quick_reactions_enabled: Option<bool>,
    pub emoji_picker_enabled: Option<bool>,
    pub language: Option<String>,

    // Sidebar settings (S6)
    pub group_unread_channels: Option<String>,
    pub limit_visible_dms_gms: Option<String>,

    // Advanced settings (S5)
    pub send_on_ctrl_enter: Option<bool>,
    pub enable_post_formatting: Option<bool>,
    pub enable_join_leave_messages: Option<bool>,
    pub enable_performance_debugging: Option<bool>,
    pub unread_scroll_position: Option<String>,
    pub sync_drafts: Option<bool>,
}

/// Status preset
#[derive(Debug, Clone, Serialize, FromRow)]
pub struct StatusPreset {
    pub id: Uuid,
    pub user_id: Option<Uuid>,
    pub emoji: String,
    pub text: String,
    pub duration_minutes: Option<i32>,
    pub is_default: bool,
    pub sort_order: i32,
    pub created_at: DateTime<Utc>,
}

/// DTO for creating status preset
#[derive(Debug, Clone, Deserialize)]
pub struct CreateStatusPreset {
    pub emoji: String,
    pub text: String,
    pub duration_minutes: Option<i32>,
}

/// Channel notification setting
#[derive(Debug, Clone, Serialize, FromRow)]
pub struct ChannelNotificationSetting {
    pub id: Uuid,
    pub user_id: Uuid,
    pub channel_id: Uuid,
    pub notify_level: String,
    pub is_muted: bool,
    pub mute_until: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// DTO for updating channel notification
#[derive(Debug, Clone, Deserialize)]
pub struct UpdateChannelNotification {
    pub notify_level: Option<String>,
    pub is_muted: Option<bool>,
    pub mute_until: Option<DateTime<Utc>>,
}

File: frontend/src/api/preferences.ts
import api from './client'

export interface UserStatus {
    text: string | null
    emoji: string | null
    expires_at: string | null
}

export interface UpdateStatusRequest {
    text?: string
    emoji?: string
    duration_minutes?: number
}

export interface StatusPreset {
    id: string
    user_id: string | null
    emoji: string
    text: string
    duration_minutes: number | null
    is_default: boolean
    sort_order: number
}

export interface UserPreferences {
    user_id: string
    notify_desktop: string
    notify_push: string
    notify_email: string
    notify_sounds: boolean
    dnd_enabled: boolean
    dnd_start_time: string | null
    dnd_end_time: string | null
    dnd_days: string
    message_display: string
    sidebar_behavior: string
    time_format: string
    mention_keywords: string[] | null
    // Display settings
    collapsed_reply_threads?: boolean
    use_military_time?: boolean
    teammate_name_display?: 'username' | 'nickname' | 'full_name'
    availability_status_visible?: boolean
    show_last_active_time?: boolean
    timezone?: string
    link_previews_enabled?: boolean
    image_previews_enabled?: boolean
    click_to_reply?: boolean
    channel_display_mode?: 'full' | 'centered'
    quick_reactions_enabled?: boolean
    emoji_picker_enabled?: boolean
    language?: string
    // Sidebar settings
    group_unread_channels?: 'never' | 'only_for_favorites' | 'always'
    limit_visible_dms_gms?: 'all' | '10' | '20' | '40'
    // Advanced settings
    send_on_ctrl_enter?: boolean
    enable_post_formatting?: boolean
    enable_join_leave_messages?: boolean
    enable_performance_debugging?: boolean
    unread_scroll_position?: 'start' | 'last' | 'end'
    sync_drafts?: boolean
}

export interface UpdatePreferencesRequest {
    notify_desktop?: string
    notify_push?: string
    notify_email?: string
    notify_sounds?: boolean
    dnd_enabled?: boolean
    dnd_start_time?: string
    dnd_end_time?: string
    dnd_days?: string
    message_display?: string
    sidebar_behavior?: string
    time_format?: string
    mention_keywords?: string[]
    // Display settings
    collapsed_reply_threads?: boolean
    use_military_time?: boolean
    teammate_name_display?: 'username' | 'nickname' | 'full_name'
    availability_status_visible?: boolean
    show_last_active_time?: boolean
    timezone?: string
    link_previews_enabled?: boolean
    image_previews_enabled?: boolean
    click_to_reply?: boolean
    channel_display_mode?: 'full' | 'centered'
    quick_reactions_enabled?: boolean
    emoji_picker_enabled?: boolean
    language?: string
    // Sidebar settings
    group_unread_channels?: 'never' | 'only_for_favorites' | 'always'
    limit_visible_dms_gms?: 'all' | '10' | '20' | '40'
    // Advanced settings
    send_on_ctrl_enter?: boolean
    enable_post_formatting?: boolean
    enable_join_leave_messages?: boolean
    enable_performance_debugging?: boolean
    unread_scroll_position?: 'start' | 'last' | 'end'
    sync_drafts?: boolean
}

export interface ChannelNotificationSetting {
    id: string
    user_id: string
    channel_id: string
    notify_level: string
    is_muted: boolean
    mute_until: string | null
}

// Mattermost-compatible preference for favorites, etc.
export interface Preference {
    user_id: string
    category: string
    name: string
    value: string
}

export const preferencesApi = {
    // User status
    getMyStatus: () => api.get<UserStatus>('/users/me/status'),
    updateMyStatus: (data: UpdateStatusRequest) => api.put<UserStatus>('/users/me/status', data),
    clearMyStatus: () => api.delete<UserStatus>('/users/me/status'),
    getUserStatus: (userId: string) => api.get<UserStatus>(`/users/${userId}/status`),

    // User preferences
    getMyPreferences: () => api.get<UserPreferences>('/users/me/preferences'),
    updateMyPreferences: (data: UpdatePreferencesRequest) => api.put<UserPreferences>('/users/me/preferences', data),

    // Status presets
    listStatusPresets: () => api.get<StatusPreset[]>('/users/me/status/presets'),
    createStatusPreset: (data: { emoji: string; text: string; duration_minutes?: number }) =>
        api.post<StatusPreset>('/users/me/status/presets', data),
    deleteStatusPreset: (presetId: string) => api.delete(`/users/me/status/presets/${presetId}`),

    // Channel notifications
    getChannelNotifications: (channelId: string) =>
        api.get<ChannelNotificationSetting | null>(`/channels/${channelId}/notifications`),
    updateChannelNotifications: (channelId: string, data: { notify_level?: string; is_muted?: boolean; mute_until?: string }) =>
        api.put<ChannelNotificationSetting>(`/channels/${channelId}/notifications`, data),
    
    // Mattermost-compatible preferences (for favorites)
    getMyPreferencesMm: () => api.get<Preference[]>('/users/me/preferences'),
    getPreferencesByCategory: (userId: string, category: string) =>
        api.get<Preference[]>(`/users/${userId}/preferences/${category}`),
    updatePreferences: (userId: string, preferences: Preference[]) =>
        api.put(`/users/${userId}/preferences`, preferences),
    deletePreferences: (userId: string, preferences: Preference[]) =>
        api.delete(`/users/${userId}/preferences`, { data: preferences }),
}
